docs/free_threaded.rst

```diff
@@ -0,0 +1,289 @@
+.. _free-threaded:
+
+.. cpp:namespace:: nanobind
+
+Free-threaded Python
+====================
+
+**Free-threading** is an experimental new Python feature that replaces the
+`Global Interpreter Lock (GIL)
+<https://en.wikipedia.org/wiki/Global_interpreter_lock>`__ with a fine-grained
+locking scheme to better leverage multi-core parallelism. The resulting
+benefits do not come for free: extensions must explicitly opt-in and generally
+require careful modifications to ensure correctness.
+
+Nanobind can target free-threaded Python since version 2.2.0. This page
+explains how to do so and discusses a few caveats. Besides this page, make sure
+to review `py-free-threading.github.io <https://py-free-threading.github.io>`__
+for a more comprehensive discussion of free-threaded Python. `PEP 703
+<https://peps.python.org/pep-0703/>`__ explains the nitty gritty details.
+
+Opting in
+---------
+
+To opt into free-threaded Python, pass the ``FREE_THREADED`` parameter to the
+:cmake:command:`nanobind_add_module()` CMake target command. For other build
+systems, refer to their respective documentation pages.
+
+.. code-block:: cmake
+
+      nanobind_add_module(
+        my_ext                   # Target name
+        FREE_THREADED            # Opt into free-threading
+        my_ext.h                 # Source code files below
+        my_ext.cpp)
```

Comment:
I'm sorry, I'm not very sure to understand the Opt-in docs part. We can opt into free-threading mode using CMake only ? For other build systems we need to define `NB_FREE_THREADED` flag?

Reply:
Cmake is the official build system of this project. You need to opt-in from there because this also changes how nanobind's add-on library is compiled. There are also a separate Bazel and Meson build systems maintained by @nicholasjng and @WillAyd. The authors of these will need to update their code with a minor adaptation, but that is not part of this PR.

When free threading is available and requested, the build sytem will automatically set `NB_FREE_THREADED` when compiling the C++ extension code. The user doesn't need to use this, but it could be useful in some cases.

Reply:
fyi, I'm tracking this in https://github.com/nicholasjng/nanobind-bazel/issues/37 for Bazel. Contributions welcome!

Reply:
Meson >= 1.5.0 natively supports building free-threaded extensions from a free-threaded interpreter. In theory, the only thing we will have to do with that configuration is add the `NB_FREE_THREADED` define when that type of interpreter is detected, although if you wanted to experiment early you can try updating your wrap configuration to point to head here and manually define the macro:

```
meson setup builddir -Dcpp_args="$CXXFLAGS -DNB_FREE_THREADED"
```

or 

```
pip install <your_extension> -Csetup-args='-Dcpp_args="$CXXFLAGS -DNB_FREE_THREADED'
```

if using the pip front-end